### PR TITLE
[Hardware][Intel CPU] Adding intel openmp tunings in Docker file

### PR DESCRIPTION
**Pull request by zhouyuan**
This patch adds more tunings for CPU backend on intel openmp. 
These tunings improves CPU backend performance greatly, especially on throughput related tests.
This patch also updates the CPU docker file to run the serving endpoint as entry point by default

FIX #xxxx (*link existing issues this PR will resolve*)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: Dockerfile.cpu [hunk 1]
```diff
@@ -6,7 +6,16 @@ RUN apt-get update  -y \
     && apt-get install -y git wget vim numactl gcc-12 g++-12 python3 python3-pip libtcmalloc-minimal4 \
     && update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-12 10 --slave /usr/bin/g++ g++ /usr/bin/g++-12
 
-RUN echo 'export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so.4:$LD_PRELOAD' >> ~/.bashrc
+RUN pip install accelerate mkl
```
Comment (WoosukKwon) [inline Dockerfile.cpu:9]: QQ: Where are these libraries used?

File: Dockerfile.cpu [hunk 2]
```diff
@@ -6,7 +6,16 @@ RUN apt-get update  -y \
...
+
+ENV LD_PRELOAD="/usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so.4:/usr/local/lib/libiomp5.so:$LD_PRELOAD"
+
+ENV KMP_BLOCKTIME=1
+ENV KMP_TPAUSE=0
+ENV KMP_SETTINGS=1
+ENV KMP_FORKJOIN_BARRIER_PATTERN=dist,dist
+ENV KMP_PLAIN_BARRIER_PATTERN=dist,dist
+ENV KMP_REDUCTION_BARRIER_PATTERN=dist,dist
```
Comment (WoosukKwon) [inline Dockerfile.cpu:18]: Could you please add comments on what these env variables are about?

File: Dockerfile.cpu (hunk 1 above)
Comment (zhouyuan) [reply Dockerfile.cpu]: I actually use `accelerate` the setup the notebook but indeed it's not necessary here - just removed the `accelerate` and `mkl` package, and install `intel-openmp` only here

File: Dockerfile.cpu (hunk 2 above)
Comment (zhouyuan) [reply Dockerfile.cpu]: Yes, added the comments and reference doc for these tunings https://github.com/vllm-project/vllm/pull/6008/commits/458db88788df56cbe2533ca82dcd7a81f8ddf587

File: Dockerfile.cpu [hunk 3]
```diff
@@ -31,4 +45,4 @@ WORKDIR /workspace/
 
 RUN ln -s /workspace/vllm/tests && ln -s /workspace/vllm/examples && ln -s /workspace/vllm/benchmarks
 
-CMD ["/bin/bash"]
+ENTRYPOINT ["python3", "-m", "vllm.entrypoints.openai.api_server"]
```
Comment (WoosukKwon) [inline Dockerfile.cpu:48]: I personally think it's ok to keep this as `/bin/bash`, but it also makes sense to keep this consistent with the CUDA dockerfile... 🤔

**Review by WoosukKwon — APPROVED** [review]
LGTM! Thanks for the PR!

File: Dockerfile.cpu [hunk 4]
```diff
@@ -6,7 +6,21 @@ RUN apt-get update  -y \
...
+ENV LD_PRELOAD="/usr/lib/x86_64-linux-gnu/libtcmalloc_minimal.so.4:/usr/local/lib/libiomp5.so:$LD_PRELOAD"
+
+# The time(milliseconds) that a thread should wait after completing the execution of a parallel region, before sleeping.
+ENV KMP_BLOCKTIME=1 
+# Prevents the CPU to run into low performance state
+ENV KMP_TPAUSE=0
+# Provides fine granularity parallelism
+ENV KMP_FORKJOIN_BARRIER_PATTERN=dist,dist
+ENV KMP_PLAIN_BARRIER_PATTERN=dist,dist
+ENV KMP_REDUCTION_BARRIER_PATTERN=dist,dist
```
Comment (WoosukKwon) [inline Dockerfile.cpu:23]: QQ: Is it possible to set these env variables inside vLLM (e.g., when the worker is initialized)?

File: Dockerfile.cpu (hunk 3 above)
Comment (zhouyuan) [reply Dockerfile.cpu:48]: was trying to fix the feature gap on issues: 
https://github.com/vllm-project/vllm/issues/4771

File: Dockerfile.cpu (hunk 4 above)
Comment (zhouyuan) [reply Dockerfile.cpu:23]: +1, yes moving to CPU worker seems better, will make the change

File: vllm/worker/cpu_worker.py [hunk 5]
```diff
@@ -150,6 +150,9 @@ def __init__(
         if self.is_driver_worker:
             assert self.rank == 0, "The driver worker must have rank 0."
 
+        # try to initialize intel openmp optimized tunings
+        init_kmp_env()
```
Comment (WoosukKwon) [inline vllm/worker/cpu_worker.py:154]: Please double-check whether the env variables are correctly used by the libraries. In my personal experience, some env variables have to be set before the libraries are loaded.

**Pull request merged**